merkletree.go

```diff
@@ -0,0 +1,344 @@
+			// add a new interior node and push the previous leaf down
+			// then continue insertion
+			currentNodeUL := nodePointer.(*userLeafNode)
+			if currentNodeUL.parent == nil {
+				return ErrInvalidTree
+			}
+			newInteriorNode := interiorNode{
+				parent: currentNodeUL.parent,
+				level:  currentNodeUL.level,
+			}
+
+			if currentNodeUL.key == node.key {
+				// replace the value
+				currentNodeUL.value = node.value
+				return nil
+			}
+
+			currentNodeKey := m.computePrivateIndex(currentNodeUL.key)
+
+			currentNodeUL.index = currentNodeKey
+
+			direction := getNthBit(currentNodeKey, position)
+			if direction {
+				newInteriorNode.rightChild = currentNodeUL
+			} else {
+				newInteriorNode.leftChild = currentNodeUL
+			}
+			currentNodeUL.level++
+			currentNodeUL.parent = &newInteriorNode
+			if newInteriorNode.parent.leftChild == nodePointer {
+				newInteriorNode.parent.leftChild = &newInteriorNode
+			} else {
+				newInteriorNode.parent.rightChild = &newInteriorNode
+			}
+			nodePointer = &newInteriorNode
+			node.level--
+		case *interiorNode:
+			currentNodeI := nodePointer.(*interiorNode)
+			direction := getNthBit(key, position)
+
+			if direction { // go right
+				currentNodeI.rightHash = nil
+				if currentNodeI.rightChild == nil {
+					currentNodeI.rightChild = node
+					node.parent = currentNodeI
+					break insertLoop
+				} else {
+					nodePointer = currentNodeI.rightChild
+				}
+			} else { // go left
+				currentNodeI.leftHash = nil
+				if currentNodeI.leftChild == nil {
+					currentNodeI.leftChild = node
+					node.parent = currentNodeI
+					break insertLoop
+				} else {
+					nodePointer = currentNodeI.leftChild
+				}
+			}
+			position++
+		default:
+			return ErrInvalidTree
+		}
+	}
+	return nil
+}
+
+func (m *MerkleTree) RecomputeHash() {
+	m.computeHash([]bool{})
+}
+
+func (m *MerkleTree) computeHash(prefixBits []bool) {
+	if m.root.leftHash == nil {
+		m.root.leftHash = m.hashNode(
+			append(prefixBits, false), m.root.leftChild)
+	}
+	if m.root.rightHash == nil {
+		m.root.rightHash = m.hashNode(
+			append(prefixBits, true), m.root.rightChild)
+	}
+}
+
+func (m *MerkleTree) hashNode(prefixBits []bool, node interface{}) []byte {
+	if node == nil {
+		// empty node
+		return hashEmptyNode(m, prefixBits)
+	}
+
+	switch node.(type) {
+	case *interiorNode:
+		currentNodeI := node.(*interiorNode)
+		if currentNodeI.leftHash == nil {
+			currentNodeI.leftHash = m.hashNode(
+				append(prefixBits, false), currentNodeI.leftChild)
+		}
+		if currentNodeI.rightHash == nil {
+			currentNodeI.rightHash = m.hashNode(
+				append(prefixBits, true), currentNodeI.rightChild)
+		}
+		return hashInteriorNode(m, currentNodeI)
+	case *userLeafNode:
+		return hashLeafNode(m, node.(*userLeafNode))
+	}
+	return nil
+}
+
+func (m *MerkleTree) digest(input []byte) []byte {
+	h := m.hash.Hash
+	defer h.Reset()
+	h.Write(input)
+	return h.Sum(nil)
+}
+
+// Node manipulation functions
```

Comment:
I see your reasoning with the method vs function naming conventions, but turning functions into methods or vice versa to have consistent naming isn't what I was thinking. I made of comment on this in a later commit, I think what would solve this may be refactoring the node types into their own module(s) for clarity. 

I think given the requirements for auth path generation, it would make sense to create an ```emptyNode`` type.